arcaflow_plugin_horreum_client/horreum_client_schema.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+
+import enum
+import typing
+import re
+from dataclasses import dataclass
+from arcaflow_plugin_sdk import schema
+
+
+class access(str, enum.Enum):
+    PUBLIC = "PUBLIC"
+    PROTECTED = "PROTECTED"
+    PRIVATE = "PRIVATE"
+
+
+@dataclass
+class InputParams:
+    horreum_url: typing.Annotated[
+        str,
+        schema.name("Horreum URL"),
+        schema.description(
+            "The complete base URL for the Horreum server,"
+            " such as 'https://horreum.example.com'."
+        ),
+        schema.pattern(
+            re.compile(
+                r"(https?:\/\/)?"
+                r"[a-zA-Z0-9_-]{1,63}(\.[a-zA-Z0-9_-]{1,63})*"
+                r"(:[0-9]{1,5})?"
+                r"\/?"
+            )
+        ),
```

Comment:
This is cute. It doesn't support a literal IPv[4|6] address, for one thing, but, fine. However, it might be worth thinking about building "robust" parsers for "common datatypes" (like URLs) into the Arca schema infrastructure rather than making each plugin craft their own. (Yeah, out of scope for this PR, but an "earbug"...)

Reply:
I completely agree that input validation for URLs should be handled by the SDK. For now, the re pattern matching is what I have to work with. I went through a couple of iterations of this (it's surprising that I don't quickly find in an Internet search some existing "perfect" pattern), and what I have here was suggested by @webbnh as an improvement to the previous more messy pattern. If I'm reading the pattern right, it should support an IPv4 address, but not an IPv6 address, but it also wouldn't allow a base URL with a sub-path, which could be a problem, too. I'm happy to review any other pattern suggestions.

Reply:
I believe that it _does_ support an IPv4 numerical address...it just doesn't handle IPv6 (and nobody uses those, right??  🙃).

It does not support a sub-path (or path prefix), but that's more or less intentional on my part, based on my reading of how the code actually uses it.

Reply:
True: an IPv4 is just "hex numeric name dot hex numeric subdomain dot hex numeric subdomain dot hex numeric domain" here; and the fact that Red Hat (and many others) still predominantly use IPv4 nested NAT domains is just incredibly ridiculous, but important customers *do* rely on IPv6, and we should never be ignoring it, but that rant is more or less orthogonal to this PR so enough. 😆

Reply:
> it's surprising that I don't quickly find in an Internet search some existing "perfect" pattern

I think that's because its a "hard" problem.  For starters, it's somewhat ill-defined (is the thing that you're trying to parse a "web address" or a "URL" or a "URI", might there be pieces of it missing, etc.); and then there's the fact that you're trying to do this with a regex, when what you need is a full grammar-based parser.  (Trying to write a regex for just IPv6 addresses is pretty hard, and then to try to place it within a web address which also has a password (which is preceded by a colon) and a port (which is preceded by a colon) and deciding whether it will be surrounded by square brackets.... 🤯)

It's almost certainly better to just call `urllib.parse.urlparse()` and let the experts do the heavy lifting (and, it'll probably be less run-time expense, too -- fancy REs get expensive fast!).

Reply:
Yeah, a complete RE that encompasses IPv6 rules would be really messy because there are so many alternatives and the rules aren't trivial.

Luckily, we're not doing something annoyingly complex like *comparing* IPv6 addresses here; and even in IPv6-only subnets most communication is likely to use DNS or DDNS names rather than literals.